spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala

```diff
@@ -0,0 +1,196 @@
+    }
+
+    // optional params should be at the end
+    procedure.parameters.sliding(2).foreach {
+      case Array(previousParam, currentParam) if previousParam.required && !currentParam.required =>
+        throw new AnalysisException("Optional parameters must be after required ones")
+      case _ =>
+    }
+  }
+
+  private def validateMethodHandle(procedure: Procedure): Unit = {
+    val params = procedure.parameters
+    val outputType = procedure.outputType
+
+    val methodHandle = procedure.methodHandle
+    val methodType = methodHandle.`type`
+    val methodReturnType = methodType.returnType
+
+    // method cannot accept var ags
+    if (methodHandle.isVarargsCollector) {
+      throw new AnalysisException("Method must have fixed arity")
+    }
+
+    // verify the number of params in the procedure match the number of params in the method
+    if (params.length != methodType.parameterCount) {
+      throw new AnalysisException("Method parameter count must match the number of procedure parameters")
+    }
+
+    // the MethodHandle API does not allow us to check the generic type
+    // so we only verify the return type is either void or iterable
+
+    if (outputType.nonEmpty && methodReturnType != classOf[java.lang.Iterable[_]]) {
+      throw new AnalysisException(
+        s"Wrong method return type: $methodReturnType; the procedure defines $outputType " +
+        "as its output so must return java.lang.Iterable of Spark Rows")
+    }
+
+    if (outputType.isEmpty && methodReturnType != classOf[Void]) {
+      throw new AnalysisException(
+        s"Wrong method return type: $methodReturnType; the procedure defines no output columns " +
+        "so must be void")
+    }
+  }
+
+  private def buildArgExprs(procedure: Procedure, args: Seq[CallArgument]): Seq[Expression] = {
+    val params = procedure.parameters
+
+    // build a map of declared parameter names to their positions
+    val nameToPositionMap = params.map(_.name).zipWithIndex.toMap
+
+    // build a map of parameter names to args
+    val nameToArgMap = buildNameToArgMap(params, args, nameToPositionMap)
+
+    // verify all required parameters are provided
+    val missingParamNames = params.filter(_.required).collect {
+      case param if !nameToArgMap.contains(param.name) => param.name
+    }
+
+    if (missingParamNames.nonEmpty) {
+      throw new AnalysisException(s"Missing required parameters: ${missingParamNames.mkString("[", ",", "]")}")
+    }
+
+    val argExprs = new Array[Expression](params.size)
+
+    nameToArgMap.foreach { case (name, arg) =>
+      val position = nameToPositionMap(name)
+      val param = params(position)
+      val paramType = param.dataType
+      val argType = arg.expr.dataType
+      if (paramType != argType) {
+        throw new AnalysisException(s"Wrong arg type for ${param.name}: expected $paramType but got $argType")
```

Comment:
This is going to force people to use value literals, like `12L`. Can we insert a cast and execute it if the type is compatible?

Reply:
I think we can try to use `Cast$canUpCast` for this.

Reply:
I added a cast if it is safe.

Reply:
All expressions will undergo optimizations so we will get rid of casts before creating an exec node.